src/libraries/System.Private.CoreLib/src/System/IO/Path.Windows.cs

```diff
@@ -208,25 +208,59 @@ static uint GetTempPathW(int bufferLen, ref char buffer)
         // name on disk.
         public static string GetTempFileName()
         {
-            var tempPathBuilder = new ValueStringBuilder(stackalloc char[PathInternal.MaxShortPath]);
-
-            GetTempPath(ref tempPathBuilder);
-
-            var builder = new ValueStringBuilder(stackalloc char[PathInternal.MaxShortPath]);
-
-            uint result = Interop.Kernel32.GetTempFileNameW(
-                ref tempPathBuilder.GetPinnableReference(), "tmp", 0, ref builder.GetPinnableReference());
-
-            tempPathBuilder.Dispose();
-
-            if (result == 0)
-                throw Win32Marshal.GetExceptionForLastWin32Error();
-
-            builder.Length = builder.RawChars.IndexOf('\0');
+            // Avoid GetTempFileNameW because it is limited to 0xFFFF possibilities, which both
+            // means that it may have to make many attempts to create the file before
+            // finding an unused name, and also that if an app "leaks" such temp files,
+            // it can prevent GetTempFileNameW succeeding at all.
+            //
+            // To make this a little more robust, generate our own name with more
+            // entropy. We could use GetRandomFileName() here, but for consistency
+            // with Unix and to retain the ".tmp" extension we will use the "tmpXXXXXX.tmp" pattern.
+            // Using 32 characters for convenience, that gives us 32^^6 ~= 10^^9 possibilities,
+            // but we'll still loop to handle the unlikely case the file already exists.
+
+            string tempPath = Path.GetTempPath();
+            byte* bytes = stackalloc byte[KeyLength];
+
+            while (true)
+            {
+                string path = string.Create(tempPath.Length + 13 /* tmpXXXXXX.tmp */, (IntPtr)bytes, (span, state) =>
```

Comment:
This will produce several time bigger code than what you would get by from simpler `string.Concat`. Is that extra code really worth it? Code size is a performance metric too.

Reply:
I was going with a suggestion above -- but I can't say it is. Will change...

Reply:
fixed.
|          Method |        Job |         Toolchain |     Mean |   Error |  StdDev | Ratio | Allocated |
|---------------- |----------- |------------------ |---------:|--------:|--------:|------:|----------:|
| GetTempFileName | Job-XHHVAN | \main\corerun.exe | 361.9 us | 3.26 us | 2.72 us |  1.00 |     112 B |
| GetTempFileName | Job-HOJECC |   \pr\corerun.exe | 334.3 us | 3.30 us | 2.92 us |  0.92 |     288 B |